protocol/x/vault/keeper/grpc_query_shares_test.go

```diff
@@ -0,0 +1,117 @@
+package keeper_test
+
+import (
+	"math/big"
+	"testing"
+
+	"github.com/cosmos/cosmos-sdk/types/query"
+	"github.com/dydxprotocol/v4-chain/protocol/dtypes"
+	testapp "github.com/dydxprotocol/v4-chain/protocol/testutil/app"
+	"github.com/dydxprotocol/v4-chain/protocol/testutil/constants"
+	vaulttypes "github.com/dydxprotocol/v4-chain/protocol/x/vault/types"
+	"github.com/stretchr/testify/require"
+)
+
+func TestOwnerShares(t *testing.T) {
+	tests := map[string]struct {
+		/* --- Setup --- */
+		// Request.
+		req *vaulttypes.QueryOwnerSharesRequest
+		// Vault ID.
+		vaultId vaulttypes.VaultId
+		// Owner shares.
+		ownerShares map[string]*big.Int
+
+		/* --- Expectations --- */
+		expectedOwnerShares []*vaulttypes.OwnerShare
+		expectedErr         string
+	}{
+		"Success": {
+			req: &vaulttypes.QueryOwnerSharesRequest{
+				Type:   vaulttypes.VaultType_VAULT_TYPE_CLOB,
+				Number: 0,
+			},
+			vaultId: constants.Vault_Clob_0,
+			ownerShares: map[string]*big.Int{
+				constants.Alice_Num0.Owner: big.NewInt(100),
+				constants.Bob_Num0.Owner:   big.NewInt(200),
+			},
+			expectedOwnerShares: []*vaulttypes.OwnerShare{
+				{
+					Owner: constants.Alice_Num0.Owner,
+					Shares: &vaulttypes.NumShares{
+						NumShares: dtypes.NewInt(100),
+					},
+				},
+				{
+					Owner: constants.Bob_Num0.Owner,
+					Shares: &vaulttypes.NumShares{
+						NumShares: dtypes.NewInt(200),
+					},
+				},
+			},
+		},
+		"Error: vault not found": {
+			req: &vaulttypes.QueryOwnerSharesRequest{
+				Type:   vaulttypes.VaultType_VAULT_TYPE_CLOB,
+				Number: 1,
+			},
+			vaultId: constants.Vault_Clob_0,
+			ownerShares: map[string]*big.Int{
+				constants.Alice_Num0.Owner: big.NewInt(100),
+				constants.Bob_Num0.Owner:   big.NewInt(200),
+			},
+			expectedErr: "vault not found",
+		},
+		"Error: nil request": {
+			req:     nil,
+			vaultId: constants.Vault_Clob_0,
+			ownerShares: map[string]*big.Int{
+				constants.Alice_Num0.Owner: big.NewInt(100),
+				constants.Bob_Num0.Owner:   big.NewInt(200),
+			},
+			expectedErr: "invalid request",
+		},
+	}
```

Comment:
What happens if a user sends a valid query, but paginates beyond the max size? 

Are there also any race conditions? For example the owners are constantly joining and leaving the vault, could a user of the query ever get an accurate full picture of all owners for a vault?

Reply:
> paginates beyond max size

then I believe all vaults will be returned

yeah query is not guaranteed to have updated information